ironfish/src/network/peers/connections/connection.ts

```diff
@@ -189,11 +207,99 @@ export abstract class Connection {
     this.send = wrapper
   }
 
-  shouldLogMessageType(messageType: string): boolean {
-    const bannedMessageTypes = [
-      InternalMessageType.peerList,
-      InternalMessageType.signal,
-    ] as string[]
+  shouldLogMessageType(messageType: NetworkMessageType): boolean {
+    const bannedMessageTypes = [NetworkMessageType.PeerList, NetworkMessageType.Signal]
     return !bannedMessageTypes.includes(messageType)
   }
+
+  parseMessage(message: Buffer): NetworkMessage {
+    let header: NetworkMessageHeader
+    try {
+      header = this.parseHeader(message)
+    } catch {
+      throw new Error(`Could not parse header from request: '${message.toString('hex')}'`)
+    }
+
+    const { body, type } = header
+
+    let messageBody: NetworkMessage
+    try {
+      messageBody = this.parseBody(header)
+    } catch {
+      const args = `(type: ${displayNetworkMessageType(type)}, body: '${body.toString('hex')}')`
+      throw new Error(`Could not parse payload from request: ${args}`)
+    }
+
+    return messageBody
+  }
+
+  private parseHeader(message: Buffer): NetworkMessageHeader {
+    let rpcId
+    let nonce
+    const br = bufio.read(message)
+    const type = br.readU8()
+    if (this.isRpcNetworkMessageType(type)) {
+      rpcId = br.readU16()
+    } else if (this.isGossipNetworkMessageType(type)) {
+      nonce = br.readBytes(16)
+    }
+    return {
+      type,
+      rpcId,
+      nonce,
+      body: br.readBytes(br.left()),
+    }
+  }
+
+  private isRpcNetworkMessageType(type: NetworkMessageType): boolean {
+    return [
+      NetworkMessageType.CannotSatisfyRequest,
+      NetworkMessageType.GetBlockHashesRequest,
+      NetworkMessageType.GetBlockHashesResponse,
+      NetworkMessageType.GetBlocksRequest,
+      NetworkMessageType.GetBlocksResponse,
+    ].includes(type)
+  }
+
+  private isGossipNetworkMessageType(type: NetworkMessageType): boolean {
+    return [NetworkMessageType.NewBlock, NetworkMessageType.NewTransaction].includes(type)
+  }
+
+  private parseBody({ rpcId, nonce, type, body }: NetworkMessageHeader): NetworkMessage {
+    switch (type) {
+      case NetworkMessageType.CannotSatisfyRequest:
+        Assert.isNotUndefined(rpcId)
+        return CannotSatisfyRequest.deserialize(rpcId)
```

Comment:
Not a blocker but something to think about. Should we match up the wrapping of serialize and deserialize functions a little better? Currently it looks like for serialize on messages that extend from multiple classes we are wrapping the serialize functionality

`NetworkMessage.serialize(RpcNetworkMessage.serialize(CannotSatisfyRequest.serialize()))` 

But for deserialize we just have a flat structure

`CannotSatisfyRequest.deserialize(rpcId)`

`rpcId` really would make more sense in the deserialize function of the `RpcNetworkMessage` and then we wouldn't have to do these Assert statements here as well

Reply:
Similar for `nonce` and `GossipNetworkMessage`

Reply:
The wrapping described isn't quite correct -- `RpcNetworkMessage.serializeWithMetadata` overrides the `NetworkMessage.serializeWithMetadata` function, so it's more like

`RpcNetworkMessage.serializeWithMetadata(CannotSatisfyRequest.serialize())`

I think there might be some opportunity for improvement in the naming maybe? Like maybe renaming `serialize` to `serializeBody` and `serializeWIthMetadata` to `serialize`. I'm open to ideas around improving the structure of that, but don't have any immediate ideas.

The problem around deserialization is that TS doesn't allow for any sort of `static abstract` functions, so it's hard to express the pattern the same way. You'll notice the serialization interface doesn't actually mention any deserialization functions.

I think one improvement we could make is moving the deserialization logic into `RpcNetworkMessage` and `GossipNetworkMessage`, I'll work on that in this PR.

Reply:
I added a PR here that attempts a refactor: https://github.com/iron-fish/ironfish/pull/1258 We could potentially just merge them separately too, since this wouldn't be a breaking change.

Reply:
Yea I definitely like that change a lot better without the Asserts.